ably/realtime_channel.go

```diff
@@ -237,14 +233,22 @@ func (c *RealtimeChannel) Attach(ctx context.Context) error {
 	}
 
 	attachTimeout := c.client.Connection.opts.realtimeRequestTimeout()
-	timeoutCtx, cancel := c.opts().contextWithTimeout(ctx, attachTimeout)
+	timeoutCtx, cancel := c.opts().contextWithTimeout(context.Background(), attachTimeout)
 	defer cancel()
-	err = res.Wait(timeoutCtx)
-	if errors.Is(err, context.DeadlineExceeded) {
-		err = newError(ErrTimeoutError, errors.New("timed out before attaching channel"))
-		c.setState(ChannelStateSuspended, err, false)
+	internalOpErr := make(chan error, 1)
+	go func() {
+		internalOpErr <- res.Wait(timeoutCtx)
+	}()
+	select {
+	case <-ctx.Done(): //Unblock channel wait, when external context is cancelled
+		return ctx.Err()
+	case err = <-internalOpErr:
+		if errors.Is(err, context.DeadlineExceeded) {
+			err = newError(ErrTimeoutError, errors.New("timed out before attaching channel"))
+			c.setState(ChannelStateSuspended, err, false)
```

Comment:
This should happen even if the user cancels their context.

In general, we should make a strong distinction between the calling user goroutine and the goroutine that performs library operations, but that's for a further refactoring that includes #184.

Reply:
I believe we agreed, user cancelling the context shouldn't affect internal library operations

Reply:
I am not getting what are you trying to convey here?

Reply:
resolved